libs/agno/agno/memory/manager.py

```diff
@@ -712,6 +717,166 @@ def _get_first_n_memories(self, user_id: str, limit: Optional[int] = None) -> Li
 
         return sorted_memories_list
 
+    def optimize_memories(
+        self,
+        user_id: Optional[str] = None,
+        strategy: Union[
+            MemoryOptimizationStrategyType, MemoryOptimizationStrategy
+        ] = MemoryOptimizationStrategyType.SUMMARIZE,
+        apply: bool = False,
+    ) -> List[UserMemory]:
+        """Optimize user memories using the specified strategy.
+
+        Args:
+            user_id: User ID to optimize memories for. Defaults to "default".
+            strategy: Optimization strategy. Can be:
+                - Enum: MemoryOptimizationStrategyType.SUMMARIZE
+                - Instance: Custom MemoryOptimizationStrategy instance
+            apply: If True, automatically replace memories in database.
+
+        Returns:
+            List of optimized UserMemory objects.
+        """
+        if user_id is None:
+            user_id = "default"
+
+        if isinstance(self.db, AsyncBaseDb):
+            raise ValueError(
+                "optimize_memories() is not supported with an async DB. Please use aoptimize_memories() instead."
+            )
+
+        # Get user memories
+        memories = self.get_user_memories(user_id=user_id)
+        if not memories:
+            log_debug("No memories to optimize")
+            return []
+
+        # Get strategy instance
+        if isinstance(strategy, MemoryOptimizationStrategyType):
+            strategy_instance = MemoryOptimizationStrategyFactory.create_strategy(strategy)
+        else:
+            # Already a strategy instance
+            strategy_instance = strategy
+
+        # Optimize memories using strategy
+        optimization_model = self.get_model()
+        optimized_memories = strategy_instance.optimize(memories=memories, model=optimization_model, user_id=user_id)
+
+        # Apply to database if requested
+        if apply:
+            log_debug(f"Applying optimized memories to database for user {user_id}")
+
+            # Collect memory_ids from optimized set
+            optimized_memory_ids = {mem.memory_id for mem in optimized_memories if mem.memory_id}
+
+            # Identify memories to delete (in original but not in optimized)
+            memories_to_delete = [
+                mem for mem in memories if mem.memory_id and mem.memory_id not in optimized_memory_ids
+            ]
+
+            # Delete removed memories
```

Comment:
Shouldn't we rather just clear all the user's memories and add the new ones? So it is a complete replace.

Reply:
This is doing the same thing. It's looping through all the user's memory and then deleting them. Do you want me to call any specific function which already pre-exists?